lib/check-meta.nix

```diff
@@ -0,0 +1,200 @@
+          ~/.config/nixpkgs/config.nix, like so:
+
+               {
+                 permittedInsecurePackages = [
+                   "${attrs.name or "«name-missing»"}"
+                 ];
+               }
+
+        '';
+
+    throwEvalHelp = { reason , errormsg ? "" }:
+      throw (''
+        Package ‘${attrs.name or "«name-missing»"}’ in ${pos_str} ${errormsg}, refusing to evaluate.
+
+        '' + ((builtins.getAttr reason remediation) attrs));
+
+    metaTypes = with lib.types; rec {
+      # These keys are documented
+      description = str;
+      longDescription = str;
+      branch = str;
+      homepage = str;
+      downloadPage = str;
+      license = either (listOf lib.types.attrs) (either lib.types.attrs str);
+      maintainers = listOf str;
+      priority = int;
+      platforms = listOf str;
+      hydraPlatforms = listOf str;
+      broken = bool;
+
+      # Weirder stuff that doesn't appear in the documentation?
+      version = str;
+      tag = str;
+      updateWalker = bool;
+      executables = listOf str;
+      outputsToInstall = listOf str;
+      position = str;
+      repositories = attrsOf str;
+      isBuildPythonPackage = platforms;
+      schedulingPriority = str;
+      downloadURLRegexp = str;
+      isFcitxEngine = bool;
+      isIbusEngine = bool;
+    };
+
+    checkMetaAttr = k: v:
+      if metaTypes?${k} then
+        if metaTypes.${k}.check v then null else "key '${k}' has a value ${v} of an invalid type ${builtins.typeOf v}; expected ${metaTypes.${k}.description}"
+      else "key '${k}' is unrecognized; expected one of: \n\t      [${lib.concatMapStringsSep ", " (x: "'${x}'") (lib.attrNames metaTypes)}]";
+    checkMeta = meta: if shouldCheckMeta then lib.remove null (lib.mapAttrsToList checkMetaAttr meta) else [];
+
+    # Check if a derivation is valid, that is whether it passes checks for
+    # e.g brokenness or license.
+    #
+    # Return { valid: Bool } and additionally
+    # { reason: String; errormsg: String } if it is not valid, where
+    # reason is one of "unfree", "blacklisted" or "broken".
+    checkValidity = attrs:
+      if hasDeniedUnfreeLicense attrs && !(hasWhitelistedLicense attrs) then
+        { valid = false; reason = "unfree"; errormsg = "has an unfree license (‘${showLicense attrs.meta.license}’)"; }
+      else if hasBlacklistedLicense attrs then
+        { valid = false; reason = "blacklisted"; errormsg = "has a blacklisted license (‘${showLicense attrs.meta.license}’)"; }
+      else if !allowBroken && attrs.meta.broken or false then
+        { valid = false; reason = "broken"; errormsg = "is marked as broken"; }
+      else if !allowBroken && attrs.meta.platforms or null != null && !lib.lists.elem platform attrs.meta.platforms then
+        { valid = false; reason = "broken"; errormsg = "is not supported on ‘${platform}’"; }
+      else if !(hasAllowedInsecure attrs) then
+        { valid = false; reason = "insecure"; errormsg = "is marked as insecure"; }
```

Comment:
Probably out of scope, but it would be a good place to reference some optional field with details about why it is marked as insecure.